tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -1978,32 +1991,68 @@ void IrEmitterUnnested::EmitTile(
   // workaround.
-  ksl->For(loop_name + "_y_in_tile",
-           /*start=*/constant(0),
-           /*end=*/
-           ceil_of_ratio(b_.CreateSub(tile_height, thread_id_info.thread_id_y),
-                         num_threads_y),
-           /*step=*/constant(1), [&](llvm::Value* y_indvar) {
-             llvm::Value* y_loc =
-                 b_.CreateAdd(thread_id_info.thread_id_y,
-                              b_.CreateMul(y_indvar, num_threads_y));
-             for (int64 j = 0; j < x_num_steps; j++) {
-               llvm::Value* x_loc =
-                   b_.CreateAdd(constant(j * step_x), start_offset_x, "x_loc");
-               IrArray::Index source_idx_x =
-                   source_idx.AddOffsetToDim(y_loc, kDimY, &b_)
-                       .AddOffsetToDim(constant(j * step_x), kDimX, &b_);
-               auto emit_element = [&] {
-                 return emit_elem_function(source_idx_x, y_loc, x_loc, j);
-               };
-               if (!x_tile_fits) {
-                 ksl->If(loop_name + "_x_in_tile",
-                         b_.CreateICmpULT(x_loc, tile_width), emit_element);
-               } else {
-                 emit_element();
-               }
-             }
-           });
+  int vector_size = mapping_scheme.GetVectorSize();
+  ksl->For(
+      loop_name + "_y_in_tile",
+      /*start=*/constant(0),
+      /*end=*/
+      ceil_of_ratio(b_.CreateSub(tile_height, thread_id_info.thread_id_y),
+                    num_threads_y),
+      /*step=*/constant(1), [&](llvm::Value* y_indvar) {
+        llvm::Value* y_loc = b_.CreateAdd(
+            thread_id_info.thread_id_y, b_.CreateMul(y_indvar, num_threads_y));
+        auto unroll = [&](bool add_if, int64 max_element, int64 vector_size) {
+          for (int64 j = 0; j < x_num_steps / vector_size; j++) {
+            // Prep some values. If we do not do this, LLVM doesn't vectorize.
+            llvm::Value* x_loc_base =
+                b_.CreateAdd(constant(j * step_x * vector_size), start_offset_x,
+                             "x_loc_base");
+            IrArray::Index source_idx_x_base =
+                source_idx.AddOffsetToDim(y_loc, kDimY, &b_)
+                    .AddOffsetToDim(constant(j * step_x * vector_size), kDimX,
+                                    &b_);
+
+            for (int i = 0; i < vector_size; i++) {
+              int old_j = j * vector_size + i;
+              llvm::Value* x_loc = x_loc_base;
+              IrArray::Index source_idx_x = source_idx_x_base;
+              if (i > 0) {
+                x_loc = b_.CreateAdd(constant(i), x_loc_base, "x_loc");
+                source_idx_x =
+                    source_idx_x_base.AddOffsetToDim(constant(i), kDimX, &b_);
+              }
+              auto emit_element = [&] {
+                return emit_elem_function(source_idx_x, y_loc, x_loc, old_j);
+              };
+              if (add_if) {
+                ksl->If(loop_name + "_x_in_tile",
```

Comment:
Is this check worth it? Do the benchmark shows that the performance has a non-negligible drop if we don't always check for tiling conditions?

Reply:
This one make a difference no. For the even row size, LLVM isn't able to remove the condition.
So LLVM doesn't vectorize at all without this.
I just disabled this path and I confirm this is needed for the even row size. Otherwise one of the new test fail.